code/modules/tgui/modules/appearance_changer.dm

```diff
@@ -0,0 +1,356 @@
+/datum/tgui_module/appearance_changer
+	name = "Appearance Editor"
+	var/flags = APPEARANCE_ALL_HAIR
+	var/mob/living/carbon/human/owner = null
+	var/obj/item/organ/external/head/head_organ = null
+	var/list/valid_species = list()
+	var/list/valid_hairstyles = list()
+	var/list/valid_facial_hairstyles = list()
+	var/list/valid_head_accessories = list()
+	var/list/valid_head_marking_styles = list()
+	var/list/valid_body_marking_styles = list()
+	var/list/valid_tail_marking_styles = list()
+	var/list/valid_body_accessories = list()
+	var/list/valid_alt_head_styles = list()
+
+	var/check_whitelist
+	var/list/whitelist
+	var/list/blacklist
+
+/datum/tgui_module/appearance_changer/New(datum/host, mob/living/carbon/human/H, check_species_whitelist = TRUE, list/species_whitelist = list(), list/species_blacklist = list())
+	..()
+	owner = H
+	head_organ = owner.get_organ("head")
+	check_whitelist = check_species_whitelist
+	whitelist = species_whitelist
+	blacklist = species_blacklist
+
+/datum/tgui_module/appearance_changer/tgui_act(action, list/params)
+	if(..())
+		return
+
+	. = TRUE
+
+	switch(action)
+
+		if("race")
+			if(can_change(APPEARANCE_RACE) && (params["race"] in valid_species))
+				var/datum/species/S = GLOB.all_species[params["race"]]
+				if(owner.set_species(S.type))
+					cut_and_generate_data()
+					// Species change creates new organs - runtimes ahoy if we forget this
+					head_organ = owner.get_organ("head")
+
+		if("gender")
+			if(can_change(APPEARANCE_GENDER))
+				if(owner.change_gender(params["gender"]))
+					cut_and_generate_data()
+
+		if("skin_tone")
+			if(can_change_skin_tone())
+				var/new_s_tone = null
+				if(owner.dna.species.bodyflags & HAS_SKIN_TONE)
+					new_s_tone = input(usr, "Choose your character's skin tone:\n(Light 1 - 220 Dark)", "Skin Tone", owner.s_tone) as num|null
+					if(isnum(new_s_tone) && (!..()))
+						new_s_tone = 35 - max(min(round(new_s_tone), 220),1)
+				else if(owner.dna.species.bodyflags & HAS_ICON_SKIN_TONE)
+					var/const/MAX_LINE_ENTRIES = 4
+					var/prompt = "Choose your character's skin tone: 1-[length(owner.dna.species.icon_skin_tones)]\n("
+					for(var/i = 1 to length(owner.dna.species.icon_skin_tones))
+						if(i > MAX_LINE_ENTRIES && !((i - 1) % MAX_LINE_ENTRIES))
+							prompt += "\n"
+						prompt += "[i] = [owner.dna.species.icon_skin_tones[i]]"
+						if(i != length(owner.dna.species.icon_skin_tones))
+							prompt += ", "
+					prompt += ")"
+
+					new_s_tone = input(usr, prompt, "Skin Tone", owner.s_tone) as num|null
+					if(isnum(new_s_tone) && (!..()))
+						new_s_tone = max(min(round(new_s_tone), length(owner.dna.species.icon_skin_tones)), 1)
+
+				if(new_s_tone)
+					owner.change_skin_tone(new_s_tone)
+
+		if("skin_color")
+			if(can_change_skin_color())
+				var/new_skin = input(usr, "Choose your character's skin colour: ", "Skin Color", owner.skin_colour) as color|null
+				if(new_skin && (!..()) && owner.change_skin_color(new_skin))
```

Comment:
Why the `!..()` here, above and below?
`..()` has already returned `FALSE` on line 29

Reply:
You can open the input, walk away, then set things by queueing up the inputs. The check for this again prevents that.